corehq/apps/sso/forms.py

```diff
@@ -0,0 +1,550 @@
+                        twbscrispy.PrependedText('is_active', ''),
+                    ),
+                    css_class="panel-body"
+                ),
+                css_class="panel panel-modern-gray panel-form-only"
+            ),
+            crispy.Div(
+                crispy.Div(
+                    crispy.Fieldset(
+                        _('Service Provider Settings'),
+                        'slug',
+                        *sp_details_form.service_provider_fields
+                    ),
+                    css_class="panel-body"
+                ),
+                css_class="panel panel-modern-gray panel-form-only"
+            ),
+            hqcrispy.FormActions(
+                twbscrispy.StrictButton(
+                    ugettext_lazy("Update Configuration"),
+                    type="submit",
+                    css_class="btn btn-primary",
+                )
+            )
+        )
+
+    def clean_slug(self):
+        slug = self.cleaned_data['slug']
+        _validate_or_raise_slugify_error(slug)
+        if IdentityProvider.objects.filter(slug=slug).exclude(id=self.idp.id).exists():
+            raise forms.ValidationError(
+                _("Another Identity Provider already exists with this slug.")
+            )
+        return slug
+
+    def clean_is_editable(self):
+        is_editable = self.cleaned_data['is_editable']
+        if is_editable:
+            _check_is_editable_requirements(self.idp)
+        return is_editable
+
+    def clean_is_active(self):
+        is_active = self.cleaned_data['is_active']
+        if is_active:
+            _check_is_editable_requirements(self.idp)
+            required_for_activation = [
+                (self.idp.entity_id, _('Entity ID')),
+                (self.idp.login_url, _('Login URL')),
+                (self.idp.logout_url, _('Logout URL')),
+                (self.idp.idp_cert_public
+                 and self.idp.date_idp_cert_expiration,
+                 _('Public IdP Signing Certificate')),
+            ]
+            not_filled_out = []
+            for requirement, name in required_for_activation:
+                if not requirement:
+                    not_filled_out.append(name)
+            if not_filled_out:
+                raise forms.ValidationError(
+                    _("Please make sure an enterprise admin has filled out "
+                      "the following details before activating SSO globally: "
+                      "{}").format(", ".join(not_filled_out))
+                )
+        return is_active
+
+    @transaction.atomic
+    def update_identity_provider(self, admin_user):
+        self.idp.name = self.cleaned_data['name']
+        self.idp.slug = self.cleaned_data['slug']
+        self.idp.is_editable = self.cleaned_data['is_editable']
+        self.idp.is_active = self.cleaned_data['is_active']
+        self.idp.last_modified_by = admin_user.username
+        self.idp.save()
+        return self.idp
```

Comment:
it's out of habit & patterns that I decorate any `update/create` methods in forms with `@transaction.atomic`. What's the downside if it is included?

Reply:
It's definitely slower, but unsure if its anything noticeable. On the SQL side, it's turning 1 `UPDATE` statement into 3 separate database calls:
```
BEGIN
UPDATE
COMMIT
```
Django might do something smart under the hood to batch the calls, but a naive implementation is going to issue 3 separate calls for those statements, so that's 3 round-trips vs 1.

[This answer](https://stackoverflow.com/questions/60166976/why-is-postgres-transaction-block-so-much-slower-and-how-to-solve/60170238#60170238) suggests that the additional log writing for small transactions significantly affects their cost, but I admit I don't know how well Postgres handles this, or the degree of the difference.

Other than that, it's mostly a conceptual cost. [The docs](https://docs.djangoproject.com/en/3.1/topics/db/transactions/#controlling-transactions-explicitly) have a bit near the bottom: "Open transactions have a performance cost for your database server. To minimize this overhead, keep your transactions as short as possible. This is especially important if you’re using atomic() in long-running processes, outside of Django’s request / response cycle."  I do see the argument that this can protect methods when future database calls are added, but I tend to treat transactions with caution, similar to locks and try/except blocks -- only the bare minimum of code lives in that block. That makes me wary of labelling non-transactional code as transactional.